rholang/src/main/k/rholang/matching-function.k

```diff
@@ -0,0 +1,206 @@
+// This is a recursive check for matching.
+rule <thread> ... <k> ... matchstdform[S:String]( S1:String P1:StringlessMatchFormSyntax ; S2:String P2:StringlessMatchFormSyntax )
+              => matchstdform[S]( S1 P1 ; S2 P2 ; S1 ==String S2 ) ... </k> ... </thread>
+
+// To account for expressions
+// Strings
+rule <thread> ... <k> ... matchstdform[S:String]( "string"[S1:String] ; "string"[S2:String] ; true )
+                    => matchstdform[S]( S1:String ; S2:String ; S1 ==String S2) ... </k> ... </thread>
+rule <thread> ... <k> ... matchstdform[S:String]( S1:String ; S2:String ; true ) => . ... </k> ... </thread>
+rule <thread> ... <k> ... matchstdform[S:String]( S1:String ; S2:String ; false ) => #(MATCHFAIL)[S] ... </k> ... </thread>
+
+// Ints
+rule <thread> ... <k> ... matchstdform[S:String]( "int"[I1:Int] ; "int"[I2:Int] ; true )
+                    => matchstdform[S]( I1:Int ; I2:Int ; I1 ==Int I2) ... </k> ... </thread>
+rule <thread> ... <k> ... matchstdform[S:String]( I1:Int ; I2:Int ; true ) => . ... </k> ... </thread>
+rule <thread> ... <k> ... matchstdform[S:String]( I1:Int ; I2:Int ; false ) => #(MATCHFAIL)[S] ... </k> ... </thread>
+
+// Bools
+rule <thread> ... <k> ... matchstdform[S:String]( "bool"[B1:Bool] ; "bool"[B2:Bool] ; true )
+                    => matchstdform[S]( B1:Bool ; B2:Bool ; B1 ==Bool B2) ... </k> ... </thread>
+rule <thread> ... <k> ... matchstdform[S:String]( B1:Bool ; B2:Bool ; true ) => . ... </k> ... </thread>
+rule <thread> ... <k> ... matchstdform[S:String]( B1:Bool ; B2:Bool ; false ) => #(MATCHFAIL)[S] ... </k> ... </thread>
+
+// As more expressions become part of rholang this will have to be updated
+
+
+
+// Terminate if the strings do not match (i.e., the two processes are not of the same syntactic sort)
+rule <thread> ... <k> ... matchstdform[S:String](S1:String P1:StringlessMatchFormSyntax ; S2:String P2:StringlessMatchFormSyntax ; false)
+              => #(MATCHFAIL)[S]  ... </k> ... </thread>
+
+rule <thread> ... <k> ... matchstdform[S:String]( E1:Exp ; E2:Exp ; false)
+              => #(MATCHFAIL)[S]  ... </k> ... </thread>
+
+// If the patterns are ever the same, as expressions or processes, we're good.
+rule <thread> ... <k> ... matchstdform[S:String]( S1:String P1:StringlessMatchFormSyntax ; S1:String P1:StringlessMatchFormSyntax ; true ) => . ... </k> ... </thread>
+rule <thread> ... <k> ... matchstdform[S:String](  E:Exp ; E:Exp ; true ) => . ... </k> ... </thread>
+
+// When the pattern is a quote "@"
+rule <thread> ... <k> ... matchstdform[S:String]( "quote"[P:HigherProc] ; "quote"[R:HigherProc] ; true)
+          => match[S]( P ; R ) ... </k> ... </thread>
+
+// When the pattern is an eval "*"
+rule <thread> ... <k> ... matchstdform[S:String]( "eval"[C:Chan] ; "eval"[D:Chan] ; true)
+          => match[S]( C ; D ) ... </k> ... </thread>
+
+// When the pattern is a "for"
+// This is not quite correct, but it's on its way. It still needs alpha equivalence, among other things.
```

Comment:
Thank you for using the apostrophe correctly!

Reply:
Incorrect apostrophes are of my pet peeves too 😂